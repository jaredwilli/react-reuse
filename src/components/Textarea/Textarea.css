
.mg-textarea {
    width: 100%;
}

.mg-textarea textarea {
    border: 1px solid #ccc;
    border-radius: 2px;
    font-size: 15px;
    padding: 12px;
    margin: 5px 0;
    outline: none;
}


