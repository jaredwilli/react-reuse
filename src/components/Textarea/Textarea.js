import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Label from '../Label';

import './Textarea.css';

/** Textarea with integrated label to enforce consistency in layout, error display, label placement, and required field marker. */
const Textarea = ({
    htmlId,
    name,
    placeholder,
    onChange,
    value,
    children,
    label,
    error,
    resize,
    width,
    height,
    required,
    ...props
}) => {

    const style = {
        borderColor: error ? 'red' : '#ccc',
        resize: resize && resize,
        width: width && width,
        height: height && height
    };

    const myClass = classNames('mg-textarea', {
        ['mg-error']: error
    });

    return (
        <div className={myClass}>
            <Label
                htmlFor={htmlId}
                label={label}
                required={required}
            />

            <textarea
                id={htmlId}
                name={name}
                placeholder={placeholder}
                onChange={onChange}
                style={style}
                value={value}
                {...props}
            >
            </textarea>

            {children}

            {error &&
                <div className={classNames({ 'mg-error-msg': error })}>
                    {error}
                </div>
            }
        </div>
    );
};

Textarea.defaultProps = {
    resize: 'none',
    required: false
};

Textarea.propTypes = {
    /** Unique HTML ID. Used for tying label to HTML input. Handy hook for automated testing. */
    htmlId: PropTypes.string.isRequired,

    /** Input name. Recommend setting this to match object's property so a single change handler can be used. */
    name: PropTypes.string.isRequired,

    /** Input label */
    label: PropTypes.string,

    /** Input type */
    type: PropTypes.oneOf(['text', 'number', 'password']),

    /** Mark label with asterisk if set to true */
    required: PropTypes.bool,

    /** Function to call onChange */
    onChange: PropTypes.func.isRequired,

    /** Placeholder to display when empty */
    placeholder: PropTypes.string,

    /** Value */
    value: PropTypes.any,

    /** Width */
    width: PropTypes.string,

    /** Height */
    height: PropTypes.string,

    /** Resize value or none */
    resize: PropTypes.string,

    /** String to display when error occurs */
    error: PropTypes.string,

    /** Child component to display next to the input */
    children: PropTypes.node
};

export default Textarea;
