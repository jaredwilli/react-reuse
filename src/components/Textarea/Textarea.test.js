import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Textarea from './Textarea';

describe('Textarea', () => {
    const props = {
        htmlId: 'test',
        name: 'Test',
        label: 'Test',
        onChange: () => true
    };

    it('renders the component correctly', () => {
        const tree = renderer.create(<Textarea {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
