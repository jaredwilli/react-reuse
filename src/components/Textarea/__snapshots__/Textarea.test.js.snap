// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea renders the component correctly 1`] = `
<div
  className="mg-textarea"
>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Test
     
  </label>
  <textarea
    id="test"
    name="Test"
    onChange={[Function]}
    placeholder={undefined}
    style={
      Object {
        "borderColor": "#ccc",
        "height": undefined,
        "resize": "none",
        "width": undefined,
      }
    }
    value={undefined}
  />
</div>
`;
